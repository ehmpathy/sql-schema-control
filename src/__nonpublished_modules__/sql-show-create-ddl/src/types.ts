export interface DatabaseConnection {
  query: (args: { sql: string; values?: (string | number)[] }) => Promise<{ rows: any[] }>;
  end: () => Promise<void>;
}
