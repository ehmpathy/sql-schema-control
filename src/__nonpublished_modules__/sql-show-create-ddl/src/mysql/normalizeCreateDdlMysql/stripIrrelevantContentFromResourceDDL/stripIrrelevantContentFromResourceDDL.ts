import { ResourceType } from '../../../../../../types';

export const stripIrrelevantContentFromResourceDDL = ({
  ddl,
  resourceType,
}: {
  ddl: string;
  resourceType: ResourceType;
}) => {
  let relevantSql = ddl;

  // if resource type is FUNCTION / PROCEDURE / VIEW, strip the definer and all else between the "CREATE" and "${RESOURCE_NAME}"
  if (
    resourceType === ResourceType.FUNCTION ||
    resourceType === ResourceType.PROCEDURE ||
    resourceType === ResourceType.VIEW
  ) {
    // e.g., "CREATE DEFINER=`root`@`%` PROCEDURE `upsert_user_description`(" -> "CREATE PROCEDURE `upsert_user_description`("
    const regex = new RegExp(`(?<=CREATE )(.*)(?=${resourceType})`);
    relevantSql = relevantSql.replace(regex, '');
  }

  // if resource type is TABLE, strip the auto increment record
  if (resourceType === ResourceType.TABLE) {
    // e.g., ") ENGINE=InnoDB AUTO_INCREMENT=2 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci" -> ") ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci"
    const regex = new RegExp('(AUTO_INCREMENT=[0-9]+ )');
    relevantSql = relevantSql.replace(regex, '');
  }

  // return relevant sql
  return relevantSql;
};
