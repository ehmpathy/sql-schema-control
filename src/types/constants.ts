export enum DefinitionType {
  RESOURCE = 'resource',
  CHANGE = 'change',
}

export enum DatabaseLanguage {
  MYSQL = 'mysql',
}

export interface DatabaseConnection {
  query: (args: { sql: string, values?: any[] }) => Promise<any>;
  end: () => Promise<void>;
}

export enum ChangeDefinitionStatus {
  NOT_APPLIED = 'NOT_APPLIED',
  UP_TO_DATE = 'UP_TO_DATE',
  OUT_OF_DATE = 'OUT_OF_DATE',
}
