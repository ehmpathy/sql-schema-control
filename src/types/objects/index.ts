export { ChangeDefinition } from './ChangeDefinition';
export { ConnectionConfig } from './ConnectionConfig';
export { ControlConfig } from './ControlConfig';
export { ControlContext } from './ControlContext';
