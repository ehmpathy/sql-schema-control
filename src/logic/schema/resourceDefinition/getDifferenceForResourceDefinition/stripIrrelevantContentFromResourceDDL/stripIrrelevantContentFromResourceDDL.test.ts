import { stripIrrelevantContentFromResourceDDL } from './stripIrrelevantContentFromResourceDDL';
import { ResourceType } from '../../../../../types';

describe('stripIrrelevantContentFromResourceDDL', () => {
  it('should strip the DEFINER from PROCEDUREs', () => {
    const relevantContent = stripIrrelevantContentFromResourceDDL({
      ddl: `
CREATE DEFINER=\`root\`@\`%\` PROCEDURE \`upsert_user_description\`(
  IN in_from_user_id BIGINT
)
BEGIN
  -- just select something, and also include a comment
  SELECT false;
END
      `.trim(),
      resourceType: ResourceType.PROCEDURE,
    });
    expect(relevantContent).not.toContain('DEFINER');
    expect(relevantContent).toMatchSnapshot();
  });
  it('should strip the DEFINER from FUNCTIONs', () => {
    const relevantContent = stripIrrelevantContentFromResourceDDL({
      ddl: `
CREATE DEFINER=\`root\`@\`%\` FUNCTION \`upsert_user_description\`(
  in_from_user_id BIGINT
) RETURNS boolean
BEGIN
  -- just select something, and also include a comment
  SELECT false;
END
      `.trim(),
      resourceType: ResourceType.FUNCTION,
    });
    expect(relevantContent).not.toContain('DEFINER');
    expect(relevantContent).toMatchSnapshot();
  });
  it('should strip the AUTO_INCREMENT index from TABLEs', () => {
    const relevantContent = stripIrrelevantContentFromResourceDDL({
      ddl: `
CREATE TABLE \`some_resource_table\` (
  \`id\` int(11) NOT NULL AUTO_INCREMENT,
  \`created_at\` datetime DEFAULT CURRENT_TIMESTAMP,
  PRIMARY KEY (\`id\`)
) ENGINE=InnoDB AUTO_INCREMENT=2 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci
      `.trim(),
      resourceType: ResourceType.TABLE,
    });
    expect(relevantContent).not.toContain('AUTO_INCREMENT=');
    expect(relevantContent).toMatchSnapshot();
  });
});
