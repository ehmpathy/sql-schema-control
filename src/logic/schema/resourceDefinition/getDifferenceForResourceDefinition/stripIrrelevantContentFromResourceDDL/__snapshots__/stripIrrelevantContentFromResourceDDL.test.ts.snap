// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stripIrrelevantContentFromResourceDDL should strip the AUTO_INCREMENT index from TABLEs 1`] = `
"CREATE TABLE \`some_resource_table\` (
  \`id\` int(11) NOT NULL AUTO_INCREMENT,
  \`created_at\` datetime DEFAULT CURRENT_TIMESTAMP,
  PRIMARY KEY (\`id\`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci"
`;

exports[`stripIrrelevantContentFromResourceDDL should strip the DEFINER from FUNCTIONs 1`] = `
"CREATE FUNCTION \`upsert_user_description\`(
  in_from_user_id BIGINT
) RETURNS boolean
BEGIN
  -- just select something, and also include a comment
  SELECT false;
END"
`;

exports[`stripIrrelevantContentFromResourceDDL should strip the DEFINER from PROCEDUREs 1`] = `
"CREATE PROCEDURE \`upsert_user_description\`(
  IN in_from_user_id BIGINT
)
BEGIN
  -- just select something, and also include a comment
  SELECT false;
END"
`;
